parser/docker_cmd.go

```diff
@@ -114,7 +120,151 @@ func (b *BuildCmd) Run() (string, error) {
 			"error":    nil,
 		},
 	})
-	return image.ID, nil
+
+	var retID = image.ID
+
+	if opts.Squash {
+		libsquash.Verbose = true
+
+		var imageReader1, pipeWriter1 = io.Pipe()
+		var imageReader2, pipeWriter2 = io.Pipe()
+		var squashedImageReader, squashedImageWriter = io.Pipe()
+
+		defer func() {
+			for _, pipeReader := range []*io.PipeReader{imageReader1, imageReader2, squashedImageReader} {
+				pipeReader.Close()
+			}
+
+			for _, pipeWriter := range []*io.PipeWriter{pipeWriter1, pipeWriter2, squashedImageWriter} {
+				pipeWriter.Close()
+			}
+		}()
+
+		// exporting in a goroutine since the io.Pipe is synchronous
+		go func() {
+			err := func() error {
+				b.reporter.Log(log.WithField("image_id", image.ID), "starting export of tar stream for squash")
+				b.reporter.Event(comm.EventOptions{
+					EventType: comm.BuildEventSquashStartExport,
+					Data: map[string]interface{}{
+						"image_id": image.ID,
+					},
+				})
+				exportOpts := docker.ExportImageOptions{
+					Name:         image.ID,
+					OutputStream: pipeWriter1,
+				}
+				if err := opts.DockerClient.Client().ExportImage(exportOpts); err != nil {
+					return err
+				}
+				b.reporter.Event(comm.EventOptions{
+					EventType: comm.BuildEventSquashFinishExport,
+					Data: map[string]interface{}{
+						"image_id": image.ID,
+					},
+				})
+				return nil
+			}()
+			if err != nil {
+				b.reporter.LogLevel(log.WithField("error", err), "error exporting image for squash", log.ErrorLevel)
+			}
+		}()
+
+		go func() {
+			err := func() error {
+				b.reporter.Log(log.WithField("image_id", image.ID), "starting export of tar stream for squash")
+				b.reporter.Event(comm.EventOptions{
+					EventType: comm.BuildEventSquashStartExport,
+					Data: map[string]interface{}{
+						"image_id": image.ID,
```

Comment:
Why does libsquash needs tar streams of the same image twice?? Is this goroutine just the same as the above but with a different pipe?